Support running graphical Factor listener from Spacemacs and reloading relevant elisp code

timor opened the pull request:
Fuel mode is responsible to connect to a Factor instance.  This change extends
the Factor layer to handle a graphical listener process, to which fuel can
connect afterwards.

A major motivation is also to make it easier to develop with different Factor
versions, which can be specified with (project-/directory-specific) variables.
When starting a Factor listener in a certain location this way, the elisp code
for fuel/factor mode is reloaded from that location.

timor commented on layers/+lang/factor/funcs.el:
```diff
@@ -8,11 +8,116 @@
 ;;; License: GPLv3
+(require 'loadhist)
+
+(defvar factor/ui-listener-root nil
```
Thanks for reviewing. Do you mean `funcs.el` or `config.el`, like below?

smile13241324 replied:
config.el for variables, funcs.el for functions, sorry being unclear there.

timor commented on CHANGELOG.develop:
```diff
@@ -1864,6 +1864,9 @@ Other:
   =evil-snipe-override-mode= is enabled (thanks to Sylvain Benner)
 **** Factor
 - Misc fixes for factor version 0.98 (thanks to timor)
+- Support running graphical Factor listener instances from Spacemacs
```
In the contributing guidelines, it says that adding my name would be optional.  Not really wanting to thank myself here, to be honest.

smile13241324 replied:
Ahhh thats ok with me, then just leave it as it is :+1:

timor commented on layers/+lang/factor/funcs.el:
```diff
@@ -8,11 +8,116 @@
       (call-interactively 'fuel-stack-effect-region)
     (call-interactively 'fuel-stack-effect-sexp)))
+
+(defun factor//fuel-elisp-dir ()
```
I found the instructions in `CONVENTIONS.org` to be confusing here:

```org
* Code guidelines
** Spacemacs core and layer
Function names follow these conventions:
- =spacemacs/xxx= is an interactive function called =xxx=
- =spacemacs//xxx= is a private function called =xxx= (implementation details)
- =spacemacs|xxx= is a /macro/ called =xxx=

Variables follow these conventions:
- =spacemacs-xxx= is a variable
- =spacemacs--xxx= is a private variable (implementation details)

** All layers
A package is initialized in a function with name =<layer>/init-xxx= where:
- =<layer>= is the layer name
- =xxx= is the package name
…
```

So does `Spacemacs core and layer` refer to every layer here (There is a section `All layers` after all, which does not contain specifics for function naming)?  I see some layers which prefix their functions in `funcs.el` as `layer/function` (e.g. `latex`), and some which prefix their functions as `spacemacs/layer-function` (e.g. `helm`), and some which seem to do both (e.g. `erc`).

smile13241324 replied:
Hmmm, basically the convention for `Code Guidelines` for `spacemacs core and layer` apply to all layers. The ones you have found need to be adapted to follow the guideline. 

I'll make a note to check if there are more and fix them in time, thanks for pointing me there.

timor replied:
@smile13241324 Since also the existing function names in this layer would have to be changed according to this convention, it might make sense to leave it like this for this PR and change all of them in a follow-up PR?

smile13241324 replied:
Sounds good

timor commented on layers/+lang/factor/funcs.el:
```diff
@@ -8,11 +8,116 @@
 ;; This file is not part of GNU Emacs.
 ;;
 ;;; License: GPLv3
+(require 'loadhist)
```
I have three functions here that use something (non-autoloadable) from `loadhist`.  I could add a `(require 'loadhist)` at the beginning of each (redundant), or add the `(require 'loadhist)` in the `:config` part of the `fuel` package init function (semantically wrong place, also breaks stand-alone file integrity of `funcs.el`).

What is the preferred way?

smile13241324 replied:
Well I have some more basic questions about this. Where does `loadhist` come from, is that part of fuel? If not how do you make sure it is already installed?

As to how and where to load it, can you say why the resource is not autoloadable? Basically you should be able to declare autoloads also from outside the package, we do this frequently with packages loaded by `use-package`. If this is really not possible then I prefer to move the require in the `:config` part of `init-fuel`. 

The reason for this is that the loading order of a layer is always funcs.el, config.el followed by package.el, so it is quite common to load `defuns` in funcs.el for which the prerequisites are supplied later by package.el. Also this follows the convention of `lazy loading everything` in spacemacs as `loadhist` is only loaded when `fuel` is finally loaded, given that `fuel` is properly autoloaded this only happens when someone really uses `fuel` the first time.

timor replied:
> Well I have some more basic questions about this. Where does `loadhist` come from, is that part of fuel? If not how do you make sure it is already installed? 

`loadhist` is part of Emacs.

Only one of the used functions from that package is autoloadable, which is `unload-feature`.  As to why that is the case, I don't know. 

It has nothing to do with fuel. The functions could be called before fuel is used, because they are used to load/unload fuel, so it feels very strange to me to couple that to the `use-package` call.

> The reason for this is that the loading order of a layer is always funcs.el, config.el followed by package.el,

If I read `LAYERS.org` correctly, it states the order is `layers.el`, `package.el`, `funcs.el`, `config.el`.  Should not make a difference here, though.

smile13241324 replied:
Ups you are right, my load order was wrong, thanks for correcting me :+1:.

Ok if this is with emacs it should still be allocated to some layer. As your layer is the only one using it it should be save to declare ownership for it by adding it to the package.el and mention it with an init function, in there you could call use-package this can then require something if necessary.

This would solve the load order issue and communicate clearly that this package is required for this layer. It also allows for later lazy loading of it when necessary.

Please have a look at the `restructuredtext` layer there the build in `rst-directives` is used in a similar way.

timor replied:
I took a look at the `rst` layer.  I think in this case it actually makes sense to add autoloads instead (like e.g. the scala layer in `funcs.el`), since I am effectively only using some emacs library functions, and have no reason to provide an initializer for `loadhist`. The `:config` section of the `fuel` initializer would also not be a good choice, since this functionality explicitly deals with loading and unloading the package at run-time.

timor commented on layers/+lang/factor/funcs.el:
```diff
@@ -8,11 +8,116 @@
+    (require 'fuel-mode)
+    (require 'factor-mode)))
+
+(defvar factor//ui-listener-process nil
```
@smile13241324 Even though it is not a configuration variable, but a state variable used by private functions only?

smile13241324 replied:
Yep, you need to mark your internal variables with the proper prefix, I think it is "--" but please check the conventions.

timor replied:
Ok, will do that.  It still feels strange because `config.el` is loaded after `funcs.el`, but contains non-user-visible state exclusively used in `funcs.el`, thus breaking file consistency.  From `LAYERS.org`:

> This file configure the layer like declaring layer variables default values and setup some other variables related to the layer.

I understand the concept of layer variables to be something that can be set by the user from `.spacemacs`.  So to me, it is actually unexpected to find internal implementation details in `config.el`.

smile13241324 replied:
Hmmm this is also not done consistently in the layers ... let me ask the other maintainers what's the official convention here.

smile13241324 replied:
Not sure when I will get an answer, for now you can leave them in funcs.el, but please move them to the top of the file then or in a function.

However please change the name to use "--" so that it is clear that it is an internal variable.

timor replied:
Will do.

smile13241324 commented on layers/+lang/factor/funcs.el:
```diff
@@ -8,11 +8,116 @@
+this path to lok for the factor executable and fuel mode
+implementation. Intended to be customized per-project.")
+
+(defvar factor/ui-listener-args ""
```
config.el for variables, funcs.el for functions, sorry being unclear there.

timor commented:
@smile13241324 I hope I have addressed all issues as per our conversation.

With the last force-push, I also added the following changes, as compared to
before
- rely on upstream (fuel) variables instead of our own as much as
  possible
- add fix to make sure that factor-mode buffers are correctly (re-)set
  up after switching the emacs-lisp fuel implementation

smile13241324 commented:
Thank you for contributing to spacemacs 💜, I have cherry picked your commit into develop.
